### DISCUSSION TITLE
[MIgration] NewRelam.Remove() is not working

### DISCUSSION THREAD

**Discussion by vardansargsyan92** [realm/realm-dotnet#2583] [Q&A]
Hello.

I have a migration function that removes "Redundant" objects coming from the old realm database into the new realm database. The problem is that when migration happens
and the NewRelam.Remove() method is called it does nothing. The objects still exist in the database.

So what I should do to somehow "SKIP" or "IGNORE" those objects?

**Comment by vardansargsyan92** [comment]
**[Update]** `NewRealm.RemoveRange()` is working fine. I guess there is an issue in `NewRealm.Remove()`'s implementation.

**Reply by nirinchev** [reply]
Can you post your code?

**Reply by vardansargsyan92** [reply]
Sure

```
private void RemoveOldGroups(Migration migration, IQueryable<dynamic> oldMasterItems, IQueryable<MasterItem> newMasterItems)
        {
            for(var i = 0; i < newMasterItems.Count(); i++)
            {
                MasterItem newMasterItem = newMasterItems.ElementAt(i);
                dynamic oldMasterItem = oldMasterItems.ElementAt(i);

                dynamic oldGroup = oldMasterItem.Group;
                var isGroupItem = (bool)oldMasterItem.IsGroupItem;
                dynamic move = oldMasterItem.Move;

                // Group Item
                if(isGroupItem && oldGroup == null)
                {
                    migration.NewRealm.Remove(newMasterItem);
                }
                else // Phantom cases
                if(move == null && oldGroup == null)
                {
                    migration.NewRealm.Remove(newMasterItem);
                }
            }
        }
```

**Reply by nirinchev** [reply]
That looks like it's going to be error prone and I wonder if something weird is happening. If we try to unwind the loop a little, let's imagine you have to remove the first `newMasterItem` (the one at index 0). The next iteration of the loop, `i` would be equal to 1. However, `newMasterItems` will now have one element less. Meaning that `newMasterItems.ElementAt(1)` would skip over the current element at position 0, that was previously at position 1. It also means that there will now be a mismatch between `oldMasterItems` and `newMasterItems`, so you'd be comparing potentially different elements. If you have primary keys assigned to your objects, it's better to look them up by primary key rather than by index as those would guarantee correspondence. You should also make sure to account for collection size changes due to removal of items to make sure you're not skipping items.

**Reply by vardansargsyan92** [reply]
@nirinchev  That makes sense, and I will check that case. Nevertheless, if that's true, the database would at least delete one item during migration (I guess it would throw an exception about collection range or something like that) but no item were deleted.

Thanks for response I'll continue investigation )))

**Reply by nirinchev** [reply]
Yeah, it's suspicious that the object is not getting deleted - I'll put together a unit test and see if I can easily repro this.
